preview/MsixCore/msixmgr/AutoPlay.cpp

```diff
@@ -0,0 +1,577 @@
+                    if (initCmdLine.Get() != nullptr)
+                    {
+                        autoPlay.initCmdLine = initCmdLine.Get();
+                    }
+
+                    //GenerateHandlerName
+                    std::wstring uniqueHandlerName;
+                    uniqueHandlerName.append(handleEvent.Get());
+
+                    std::wstring generatedHandlerName;
+                    RETURN_IF_FAILED(GenerateHandlerName(L"DesktopAppXDevice", uniqueHandlerName.c_str(), generatedHandlerName));
+                    autoPlay.generatedhandlerName = generatedHandlerName.c_str();
+
+                    m_autoPlay.push_back(autoPlay);
+                    RETURN_IF_FAILED(desktopAppxDeviceEnum->MoveNext(&has_DesktopAppxDevice));
+
+                }
+
+                RETURN_IF_FAILED(invokeActionEnum->MoveNext(&hc_invokeAction));
+            }
+        }
+
+        RETURN_IF_FAILED(extensionEnum->MoveNext(&hasCurrent));
+    }
+
+    return S_OK;
+}
+
+HRESULT AutoPlay::GenerateProgId(std::wstring categoryName, std::wstring subCategory, std::wstring & generatedProgId)
+{
+    std::wstring packageMoniker = m_msixRequest->GetPackageInfo()->GetPackageFamilyName();
+    std::wstring applicationId = m_msixRequest->GetPackageInfo()->GetApplicationId();
+
+    if (packageMoniker.empty() || applicationId.empty() || categoryName.empty())
+    {
+        return E_INVALIDARG;
+    }
+
+    // Constants
+    std::wstring AppXPrefix = L"AppX";
+    static const size_t MaxProgIDLength = 39;
+
+    // The maximum number of characters we can have as base32 encoded is 32.
+    // We arrive at this due to the interface presented by the GetChars function;
+    // it is byte only. Ideally, the MaxBase32EncodedStringLength would be
+    // 35 [39 total for the progID, minus 4 for the prefix]. 35 characters means
+    // a maximum of 22 bytes in the digest [ceil(35*5/8)]. However, 22 bytes of digest
+    // actually encodes 36 characters [ceil(22/8*5)]. So we have to reduce the
+    // character count of the encoded string. At 33 characters, this translates into
+    // a 21 byte buffer. A 21 byte buffer translates into 34 characters. Although
+    // this meets the requirements, it is confusing since a 33 character limit
+    // results in a 34 character long encoding. In comparison, a 32 character long
+    // encoding divides evenly and always results in a 20 byte digest.
+
+    static const size_t MaxBase32EncodedStringLength = 32;
+
+    // The maximum number of bytes the digest can be is 20. We arrive at this by:
+    // - The maximum count of characters [without prefix] in the encoding (32):
+    // - multiplied by 5 (since each character in base 32 encoding is based off of 5 bits)
+    // - divided by 8 (to find how many bytes are required)
+    // - The  initial plus 7 is to enable integer math (equivalent of writing ceil)
+
+    static const ULONG MaxByteCountOfDigest = (MaxBase32EncodedStringLength * 5 + 7) / 8;
+
+    HRESULT hr = S_OK;

```

Comment:
>hr [](start = 12, length = 2)

is this used?

Reply:
No, the method just returns S_OK at the end and there are RETURN_IF_FAILED in between if a failure occurs.